system/jlib/jsocket.cpp

```diff
@@ -368,6 +368,77 @@ enum SOCKETMODE { sm_tcp_server, sm_tcp, sm_udp_server, sm_udp, sm_multicast_ser
 # endif
 #endif
 
+static CriticalSection queryKACS;
+
+enum UseKA { UNINIT, DISABLED, ENABLED };
+static std::atomic<UseKA> doKeepAlive { UNINIT };
+static int keepAliveTime = -1;
+static int keepAliveInterval = -1;
+static int keepAliveProbes = -1;
+
+/*
+<Software>
+  <Globals>
+    <expert>
+      <keepalive time="200" interval="75" probes="9"/>
+    </expert>
+  </Globals>
+
+global:
+  expert:
+    keepalive:
+      time: 200
+      interval: 75
+      probes: 9
+*/
+
+static bool queryKeepAlive(int &time, int &intvl, int &probes)
+{
+    if (doKeepAlive == UNINIT)
+    {
+        CriticalBlock block(queryKACS);
+
+        if (doKeepAlive == UNINIT)
+        {
+            doKeepAlive = DISABLED;
+
+#ifdef _CONTAINERIZED
+            // NOTE: toposerver does not load globals
+            Owned<IPropertyTree> expert = getGlobalConfigSP()->getPropTree("expert");
+#else
+            // MCK - without this many components will not have a global prop
+            IPropertyTree *expert = nullptr;
+            Owned<IPropertyTree> envtree = getHPCCEnvironment();
+            if (envtree)
+                expert = envtree->queryPropTree("Software/Globals/expert");
+#endif
+            if (expert)
+            {
+                IPropertyTree *keepalive = expert->queryPropTree("keepalive");
+                if (keepalive)
+                {
+                    doKeepAlive = ENABLED;
```

Comment:
also too early, another thread() could enter queryKeepAlive, just after this thread sets doKeepAlive  to ENABLED, and thefore the other thread(s) will skip the block and set the reference params to the yet as unset keepAliveTime, keepAliveInterval and keepAliveProbes

Reply:
Yes, agree.  Will fix.

Reply:
I could hold the CS always, or CriticalUnblock ub(queryKACS) after the keepalive check.  Which do you think is better ?